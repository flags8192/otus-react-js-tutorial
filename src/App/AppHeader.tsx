import React, { FC, memo, useContext } from 'react';
import Panel from '@src/layout/Panel';
import styled from '@emotion/styled';
import Button from '@src/components/Button';
import store from '@src/store';
import { resetApp, setLevel } from '@src/App/actions';
import Select from '@src/components/Select';

const Title = styled.p`
  margin: auto 20px;
`;

const AppHeader: FC = () => {
  const { dispatch } = useContext(store);

  const handleReset = () => {
    dispatch(resetApp());
  };

  const handleSelectLevel = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const { value } = event.currentTarget;
    dispatch(setLevel(value));
  };

  return (
    <Panel role="topPanel">
      <Title>Lines</Title>
      <Button onClick={handleReset} title="Reset" />
      <Select onSelect={handleSelectLevel} />
    </Panel>
  );
};

export default memo(AppHeader);
