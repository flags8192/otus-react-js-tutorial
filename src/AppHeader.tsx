import React, { PureComponent } from 'react';
import styled from '@emotion/styled';
import Input from '@src/common/Input';

const Root = styled.div`
  height: 20vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
`;

const Title = styled.div`
  margin: 10px;
`;

interface IAppHeader {
  onChangeFilter: (val: string) => void;
}
export class AppHeader extends PureComponent<IAppHeader> {
  constructor(props: IAppHeader) {
    super(props);
  }

  render() {
    return (
      <Root>
        <Title role="headerTitle">Дз №4</Title>
        <Input onChange={this.props.onChangeFilter} autoFocus={true} />
      </Root>
    );
  }
}
