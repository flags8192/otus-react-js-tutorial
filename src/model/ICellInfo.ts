enum ICellInfo {
  alive = 1,
  dead = 0,
}

export default ICellInfo;
