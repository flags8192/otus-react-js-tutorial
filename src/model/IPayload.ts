interface IPayload {
  gameFieldSize?: number;
  selected?: boolean;
  cellId?: number;
  gameLevel?: string;
}

export default IPayload;
