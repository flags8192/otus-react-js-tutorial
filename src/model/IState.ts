import ICellInfo from '@src/model/ICellInfo';
import { ISelectedCells } from '@src/modules/Cell/models';

export interface IState {
  gameFieldSize: number;
  gameFieldPercentFilled: number;
  gameFieldData: Array<ICellInfo>;
  gameLevel: string;
  selectedCells: ISelectedCells;
}
