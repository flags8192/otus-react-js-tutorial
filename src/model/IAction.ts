import IPayload from './IPayload';

export interface IAction {
  type: string;
  payload?: IPayload;
}
