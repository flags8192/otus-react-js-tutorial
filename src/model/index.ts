import { IState } from '@src/model/IState';
import { IAction } from '@src/model/IAction';

export { IAction, IState };
