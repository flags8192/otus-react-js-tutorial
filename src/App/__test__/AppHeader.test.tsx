import React from 'react';
import { render, screen } from '@testing-library/react';
import AppHeader from '@src/App/AppHeader';
import Store from '@src/store';
import { IAction } from '@src/model';

let dispatch: (action: IAction) => void;

describe('AppHeader', () => {
  beforeEach(() => {
    dispatch = jest.fn();
  });

  it('Render <AppHeader>', () => {
    const state = {
      gameLevel: '1',
      gameFieldSize: 3,
      gameFieldPercentFilled: 10,
      gameFieldData: [1, 0, 0, 0, 0, 0, 0, 0, 0],
      selectedCells: {},
    };
    const { asFragment } = render(
      <Store.Provider value={{ dispatch, state }}>
        <AppHeader />
      </Store.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const title = screen.getByText(/Lines/gi);
    expect(title).toBeInTheDocument();
    const button = screen.getByRole(/button/gi);
    expect(button).toBeInTheDocument();
    const select = screen.getByRole(/select/gi);
    expect(select).toBeInTheDocument();
  });
});
