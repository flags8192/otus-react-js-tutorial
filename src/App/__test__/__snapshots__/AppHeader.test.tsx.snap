// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppHeader Render <AppHeader> 1`] = `
<DocumentFragment>
  <div
    class="css-cogpxn"
    role="topPanel"
  >
    <p
      class="css-o9gtt2"
    >
      Lines
    </p>
    <button
      class="css-tv5zn1"
      role="button"
    >
      Reset
    </button>
    <select
      class="css-tv5zn1"
      name="select"
      role="select"
    >
      <option
        role="option"
        value="1"
      >
        Падаван
      </option>
      <option
        role="option"
        value="2"
      >
        Джедай
      </option>
      <option
        role="option"
        value="3"
      >
        Мастер-джедай
      </option>
      <option
        role="option"
        value="4"
      >
        Член Совета Силы
      </option>
    </select>
  </div>
</DocumentFragment>
`;
